### PR TITLE
Add support for submitting/receiving CRD objects as yaml

### LINKED ISSUE #37455: ThirdPartyResources / API: Failure on POST with YAML
**Is this a BUG REPORT or FEATURE REQUEST?**: Bug Report.
**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.5", GitCommit:"5a0a696437ad35c133c0c8493f7e9d22b0f9b81b", GitTreeState:"clean", BuildDate:"2016-10-29T01:38:40Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.5+coreos.0", GitCommit:"f70c2e5b2944cb5d622621a706bdec3d8a5a9c5e", GitTreeState:"clean", BuildDate:"2016-10-31T19:16:47Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**What happened**:

Occurs only with third-party resources:

```
$ curl --silent -H "Content-Type: application/yaml" -XPOST -d"$(cat manifest.yaml)" "http://127.0.0.1:8080/apis/<third_party_domain>/v1/namespaces/<namespace>/<third_party_resource_type>"
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "the object provided is unrecognized (must be of type ThirdPartyResourceData): invalid character 'a' looking for beginning of value (61706956657273696f6e3a20746563746f6e69632d76657273696f6e2e74 ...)",
  "reason": "BadRequest",
  "code": 400
}
```

The 'a' being the first letter of the manifest (stands for `apiVersion`). Changing it modifies the returned error.

However, it works just fine after conversion from YAML to JSON of the object:
```
$ curl --silent -H "Content-Type: application/json" -XPOST -d"$(cat manifest.json)" "http://127.0.0.1:8080/apis/<third_party_domain>/v1/namespaces/<namespace>/<third_party_resource_type>"
```

Of course, when passing an invalid format, the API Server answers that both JSON and YAML are supported:
```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "the body of the request was in an unknown format - accepted media types include: application/json, application/yaml",
  "reason": "UnsupportedMediaType",
  "code": 415
}
```

It also works wonderfully using `kubectl create -f manifest.yaml` because the YAML manifest is converted client-side to JSON before calling out to the API server.

**What you expected to happen**:

The adequate serializer should have been selected and should have converted the object server-side, as specified by the `Content-Type` header.

### PR DESCRIPTION
**Pull request by liggitt**
Fixes #37455

```release-note
Custom resources can now be submitted to and received from the API server in application/yaml format, consistent with other API resources.
```

### CODE REVIEW

**Comment by liggitt** [comment]
Updated with integration test:
* ensures create/update/delete/get/list accept/return YAML
* ensures patch/watch reject YAML
* ensures error cases return v1 Status objects

**Comment by liggitt** [comment]
cc @kubernetes/sig-api-machinery-pr-reviews

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 1]
```diff
@@ -32,17 +32,34 @@ import (
...
 // PatchResource returns a function that will handle a resource patch
 // TODO: Eventually PatchResource should just use GuaranteedUpdate and this routine should be a bit cleaner
-func PatchResource(r rest.Patcher, scope RequestScope, admit admission.Interface, converter runtime.ObjectConvertor) http.HandlerFunc {
+func PatchResource(r rest.Patcher, scope RequestScope, admit admission.Interface, converter runtime.ObjectConvertor, patchTypes []string) http.HandlerFunc {
 	return func(w http.ResponseWriter, req *http.Request) {
+		// Do this first, otherwise name extraction can fail for unrecognized content types
+		// TODO: handle this in negotiation
+		contentType := req.Header.Get("Content-Type")
+		// Remove "; charset=" if included in header.
+		if idx := strings.Index(contentType, ";"); idx > 0 {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: nit: >= 0

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 2]
```diff
@@ -32,17 +32,34 @@ import (
...
 	"k8s.io/apiserver/pkg/registry/rest"
 )
 
 // PatchResource returns a function that will handle a resource patch
 // TODO: Eventually PatchResource should just use GuaranteedUpdate and this routine should be a bit cleaner
-func PatchResource(r rest.Patcher, scope RequestScope, admit admission.Interface, converter runtime.ObjectConvertor) http.HandlerFunc {
+func PatchResource(r rest.Patcher, scope RequestScope, admit admission.Interface, converter runtime.ObjectConvertor, patchTypes []string) http.HandlerFunc {
 	return func(w http.ResponseWriter, req *http.Request) {
+		// Do this first, otherwise name extraction can fail for unrecognized content types
+		// TODO: handle this in negotiation
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: what's meant with the TODO? (I know it pre-existed)

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: Preexisting, and I don't think so… leading ; isn't a valid mime type attribute

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: It means to switch this to follow a similar flow with the negotiating serializer, that's just not built for dealing with patch mime types as-is

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: was reading "in negation". negotiation makes more sense actually.

**Comment by sttts** [comment]
Looks good.

**Comment by jennybuckley** [comment]
@lavalamp

**Pull request merged**